samza-core/src/main/java/org/apache/samza/storage/TransactionalStateTaskRestoreManager.java

```diff
@@ -0,0 +1,525 @@
+              checkpointDirs.forEach(checkpointDir -> storeDirsToDelete.put(storeName, checkpointDir)));
+          storesToRestore.put(storeName, new RestoreOffsets(oldestOffset, newestOffset));
+        } else { // happy path. checkpointed offset is in range of current oldest and newest offsets
+          if (!checkpointDirsOptional.isPresent()) { // non-persistent logged store
+            storesToRestore.put(storeName, new RestoreOffsets(oldestOffset, checkpointedOffset));
+          } else { // persistent logged store
+            // if there exists a valid store checkpoint directory with oldest offset <= local offset <= checkpointed offset,
+            // retain it and restore the delta. delete all other checkpoint directories for the store. if more than one such
+            // checkpoint directory exists, retain the one with the highest local offset and delete the rest.
+            boolean hasValidCheckpointDir = false;
+            for (File checkpointDir: checkpointDirsOptional.get()) {
+              // TODO BLOCKER pmaheshw: should validation check / warn for compact lag config staleness too?
+              if (storageManagerUtil.isLoggedStoreValid(
+                  storeName, checkpointDir, config, storeChangelogs, taskModel, clock, storeEngines)) {
+                String localOffset = storageManagerUtil.readOffsetFile(
+                    checkpointDir, Collections.singleton(changelogSSP), false).get(changelogSSP);
+                LOG.info("Read local offset: {} for store: {} checkpoint dir: {} in task: {}", localOffset, storeName,
+                    checkpointDir, taskName);
+
+                if (admin.offsetComparator(localOffset, oldestOffset) >= 0 &&
+                    admin.offsetComparator(localOffset, checkpointedOffset) <= 0 &&
+                    (storesToRestore.get(storeName) == null ||
+                        admin.offsetComparator(localOffset, storesToRestore.get(storeName).startingOffset) > 0)) {
+                  hasValidCheckpointDir = true;
+                  storeDirToRetain.put(storeName, checkpointDir);
+                  LOG.info("Temporarily retaining checkpoint dir: {}", checkpointDir);
+                  // mark for restore even if local == checkpointed, so that the changelog gets trimmed.
+                  storesToRestore.put(storeName, new RestoreOffsets(localOffset, checkpointedOffset));
+                }
+              }
+            }
+
+            // delete all non-retained checkpoint directories
+            for (File checkpointDir: checkpointDirsOptional.get()) {
+              if (storeDirToRetain.get(storeName) == null ||
+                  !storeDirToRetain.get(storeName).equals(checkpointDir)) {
+                storeDirsToDelete.put(storeName, checkpointDir);
+              }
+            }
+
+            // if the store had not valid checkpoint dirs to retain, restore from changelog
+            if (!hasValidCheckpointDir) {
```

Comment:
So for the job which just turned on this feature, there will not be any checkpoint store folders. Then what's the behavior? Will it bootstrap the changelog?

Reply:
As discussed offline, we will make the following changes:
1. In the next release, stores will create the checkpoint directories and write the changelog offset in the checkpoint regardless of whether the transactional state feature is turned on or not.
2. We will recommend that users first upgrade and deploy the new version with the feature turned off, and then after some time (at least one commit), turn the feature on.
This way we will have a checkpointed changelog offset + a store checkpoint directory available when the feature is turned on.
3. We will backport the change to handle null changelog offsets in the checkpoint (in CheckpointSerde) to version 1.2, to support rolling back to a previous version.